File: receipt.py
"""
Receipt-Bearing Protocol (RBP) — 跨 vendor agent 交接驗證協議 v0.1
純標準庫實現，無第三方依賴。

約定：
- 交接工作區 = 本檔同層嘅 receipts/ 目錄（outputs 同 receipt 都放呢度；v0.2 會抽做可配置 base_dir）
- receipt 嘅 outputs 字段 = map：filename -> {content_digest, canonical_digest, status}

v0.1 已知取捨（待 v0.2）：
- 驗簽（DID/VC）略過，只驗 digest
- per-output status 用 PASS/REJECT（§6.2 空輸出四態係 v0.2 增強）
- from/to.vendor 預設 fallback 做 agent_id（caller 應傳真實 vendor）
"""

import hashlib
import json
import re
import time
from pathlib import Path

# ── typed_reason 詞表 ──────────────────────────────────────────────
PASS = "PASS"
REJECT = "REJECT"
UNKNOWN = "UNKNOWN"
QUARANTINE = "QUARANTINE"

NOT_CHECKED = "NOT_CHECKED"
INDETERMINATE = "INDETERMINATE"
CONFIRMED_UNAVAILABLE = "CONFIRMED_UNAVAILABLE"
CHECKED_EMPTY = "CHECKED_EMPTY"

UNSET = "UNSET"
EMPTY = "EMPTY"

REQUIRED_FIELDS = [
    "protocol", "version", "task_id",
    "from", "to", "epoch",
    "typed_reason", "outputs",
]
MIN_DENSITY = 0.6

# ── 目錄 ────────────────────────────────────────────────────────────
_receipts_dir = Path(__file__).parent / "receipts"   # 交接工作區（outputs + receipt）
_receipts_dir.mkdir(exist_ok=True)


# ── canonical_json：RFC 8785 JCS ───────────────────────────────────
def canonical_json(obj) -> str:
    """JCS canonical JSON：鍵按 UTF-8 排序、無多餘空白。"""
    return json.dumps(
        obj,
        sort_keys=True,
        separators=(",", ":"),
        ensure_ascii=True,
        allow_nan=False,
    )


# ── digest 工具 ─────────────────────────────────────────────────────
def digest_bytes(data: bytes) -> str:
    """SHA-256 hex digest，前綴 'sha256:'。"""
    return "sha256:" + hashlib.sha256(data).hexdigest()


def content_digest(path: Path) -> str:
    """對檔案原始 bytes 做 digest_bytes。"""
    return digest_bytes(path.read_bytes())


def canonical_digest(obj) -> str:
    """對 canonical_json(obj) 的 UTF-8 bytes 做 digest_bytes。"""
    return digest_bytes(canonical_json(obj).encode("utf-8"))


def canonical_digest_file(path: Path) -> str:
    """對檔案做 canonical digest：
    - JSON object → JCS + SHA-256（消解「語義相同、字節不同」）
    - 其他（含非 JSON、JSON list/int/str）→ fall back 做 content_digest（raw bytes）
    （負例 NON-JSON-CANONICALIZE：對非 JSON 做 json.loads 會 crash；JSON list 走錯路徑。）"""
    raw = path.read_bytes()
    try:
        obj = json.loads(raw)
        if not isinstance(obj, dict):
            raise ValueError("not a JSON object")
    except (json.JSONDecodeError, ValueError):
        return digest_bytes(raw)  # fall back 做 content_digest
    return canonical_digest(obj)


# ── 路徑／ID 安全 ───────────────────────────────────────────────────
def _safe_name(name) -> bool:
    """只接受單一相對 filename（拒 absolute／`..`／path separator／NUL／首尾空白）。
    （負例：path traversal 逃逸 receipts/ 目錄。）"""
    if not isinstance(name, str) or not name:
        return False
    if name != name.strip() or "\x00" in name:
        return False
    if name in (".", "..") or ".." in name:
        return False
    if "/" in name or "\\" in name:
        return False
    if name.startswith("/") or name.startswith("\\"):
        return False
    return True


def _safe_task_id(task_id) -> bool:
    """task_id 只允許 [A-Za-z0-9_-]{1,128}，防寫出 receipts/ 目錄。"""
    return isinstance(task_id, str) and bool(re.fullmatch(r"[A-Za-z0-9_-]{1,128}", task_id))


# ── write_receipt ──────────────────────────────────────────────────
def write_receipt(
    task_id: str,
    from_agent: str,
    to_agent: str,
    prev_outputs: list,
    new_outputs: list,
    from_vendor: str = None,
    to_vendor: str = None,
    typed_reason: str = PASS,
    note: str = "",
) -> dict:
    """
    構建 receipt dict 並寫入 receipts/{task_id}.json。
    - outputs 累積 prev_outputs + new_outputs，落成 map：filename -> {content_digest, canonical_digest, status}
    - 空 output（0 bytes）自動降級 REJECT（fail-closed）
    - 空 path（冇產出）略過，唔落 map
    prev_outputs / new_outputs 每項係 str filename 或 dict {path, status}。
    """
    if not _safe_task_id(task_id):
        raise ValueError(f"invalid task_id: {task_id!r}")

    outputs: dict = {}   # filename -> {content_digest, canonical_digest, status}
    final_reason = typed_reason

    for src in (prev_outputs, new_outputs):
        for item in src:
            if isinstance(item, dict):
                path_str = item.get("path", "")
            else:
                path_str = str(item)

            if not path_str:
                continue   # 空 path = 冇產出，略過

            if not _safe_name(path_str):
                final_reason = REJECT   # 唔安全 path（../ 等）fail-closed
                continue

            p = _receipts_dir / path_str
            if not p.exists():
                final_reason = REJECT
                outputs[path_str] = {
                    "content_digest": digest_bytes(b""),
                    "canonical_digest": digest_bytes(b""),
                    "status": REJECT,
                }
                continue

            raw_bytes = p.read_bytes()
            if len(raw_bytes) == 0:
                final_reason = REJECT   # 空輸出 = silent failure，fail-closed
                outputs[path_str] = {
                    "content_digest": digest_bytes(b""),
                    "canonical_digest": digest_bytes(b""),
                    "status": REJECT,
                }
                continue

            outputs[path_str] = {
                "content_digest": content_digest(p),
                "canonical_digest": canonical_digest_file(p),
                "status": PASS,
            }

    receipt = {
        "protocol": "rbp",
        "version": "0.1",
        "task_id": task_id,
        "from": {"agent_id": from_agent, "vendor": from_vendor or from_agent},
        "to": {"agent_id": to_agent, "vendor": to_vendor or to_agent},
        "epoch": int(time.time()),
        "typed_reason": final_reason,
        "note": note,
        "outputs": outputs,
    }

    out_path = _receipts_dir / f"{task_id}.json"
    out_path.write_text(canonical_json(receipt), encoding="utf-8")
    return receipt


# ── verify_receipt ─────────────────────────────────────────────────
def verify_receipt(
    task_id: str,
    expect_from: str,
    max_age: int = 3600,
) -> tuple:
    """
    五步驗證，順序：density → scope → digest → epoch → verdict。
    任何一步唔過回傳 (False, 原因字串)，fail-closed。
    ① 驗簽（v0.1 略過）→ 由 scope 做身份前置校驗，再驗 digest（防 scope 被 digest 掩蓋）。
    """
    path = _receipts_dir / f"{task_id}.json"

    if not path.exists():
        return False, f"receipt file not found: {path}"

    try:
        receipt = json.loads(path.read_bytes())
    except (OSError, json.JSONDecodeError) as exc:
        return False, f"failed to read receipt: {exc}"

    if not isinstance(receipt, dict):
        return False, "receipt is not a JSON object"

    # ── ① density（deserialization 之後、semantic verification 之前）──
    ok, density = density_check(receipt)
    if not ok:
        return False, f"INCOMPLETE: density {density:.2f} < {MIN_DENSITY}"

    # ── ② scope：身份前置校驗（v0.1 略過驗簽，scope 緊接 density 之後）──
    sender = receipt.get("from", {})
    if not isinstance(sender, dict) or sender.get("agent_id") != expect_from:
        return False, f"scope mismatch: receipt.from={sender}, expected={expect_from}"

    # ── ③ digest：對 outputs 每個檔重算 ─────────────────────────────
    outputs = receipt.get("outputs", {})
    if not isinstance(outputs, dict):
        return False, "outputs is not a map"
    for name, meta in outputs.items():
        if not _safe_name(name):
            return False, f"unsafe output name: {name!r}"
        if not isinstance(meta, dict):
            return False, f"output metadata not a map for {name!r}"
        p = _receipts_dir / name
        if not p.is_file():
            return False, f"output file missing: {name}"
        try:
            if content_digest(p) != meta.get("content_digest"):
                return False, f"content_digest mismatch for {name}"
            if canonical_digest_file(p) != meta.get("canonical_digest", ""):
                return False, f"canonical_digest mismatch for {name}"
        except (OSError, ValueError, TypeError, RecursionError) as exc:
            return False, f"digest failed for {name}: {exc}"

    # ── ④ epoch：本地時鐘，防舊 state 同未來時間 ────────────────────
    epoch = receipt.get("epoch", 0)
    if not isinstance(epoch, int) or isinstance(epoch, bool):
        return False, f"invalid epoch type: {type(epoch).__name__}"
    age = int(time.time()) - epoch
    if age < 0 or age > max_age:
        return False, f"epoch invalid/stale: age={age}s (max_age={max_age}s)"

    # ── ⑤ verdict ────────────────────────────────────────────────────
    if receipt.get("typed_reason") != PASS:
        return False, f"verdict not PASS: {receipt.get('typed_reason')}"

    return True, "OK"


# ── density_check ──────────────────────────────────────────────────
def density_check(receipt: dict) -> tuple:
    """
    計算 receipt 中 required fields 的覆蓋密度 + outputs 非空。
    回傳 (通過?, metric)。
    （負例 HOLLOW-RECEIPT：valid JSON 但內容空洞，outputs 空 → 不過。）"""
    present = sum(
        1 for field in REQUIRED_FIELDS
        if field in receipt and receipt[field] is not None
    )
    metric = present / len(REQUIRED_FIELDS)
    if metric < MIN_DENSITY:
        return False, metric
    outputs = receipt.get("outputs")
    if not isinstance(outputs, dict) or len(outputs) == 0:
        return False, 0.0   # 空 outputs = hollow receipt，fail
    return True, metric

File: receipt.ts
/**
 * Receipt-Bearing Protocol (RBP) — TypeScript (Node.js) reference implementation v0.1
 * Pure standard library (node:crypto, node:fs, node:path), zero third-party deps.
 *
 * The one hard interop requirement is canonical digest:
 *   sha256(JCS(value))   where JCS = RFC 8785 JSON Canonicalization Scheme.
 * Conformance: reproduce the golden vectors in fixtures/jcs-vectors.json.
 *
 * Known JS limitation: JSON integers beyond 2^53 lose precision when parsed
 * with JSON.parse, so the `numbers` golden vector's `big` field cannot be
 * reproduced byte-for-byte in JS (Python preserves it). Every other vector
 * passes. See fixtures/README.md.
 */

import { createHash } from 'node:crypto';
import { existsSync, mkdirSync, readFileSync, writeFileSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';

// ── typed_reason 詞表 ──────────────────────────────────────────────
export const PASS = 'PASS';
export const REJECT = 'REJECT';
export const UNKNOWN = 'UNKNOWN';
export const QUARANTINE = 'QUARANTINE';

export const NOT_CHECKED = 'NOT_CHECKED';
export const INDETERMINATE = 'INDETERMINATE';
export const CONFIRMED_UNAVAILABLE = 'CONFIRMED_UNAVAILABLE';
export const CHECKED_EMPTY = 'CHECKED_EMPTY';

export const UNSET = 'UNSET';
export const EMPTY = 'EMPTY';

export const REQUIRED_FIELDS = [
  'protocol', 'version', 'task_id',
  'from', 'to', 'epoch',
  'typed_reason', 'outputs',
];
export const MIN_DENSITY = 0.6;

// ── 目錄 ────────────────────────────────────────────────────────────
const receiptsDir = join(dirname(fileURLToPath(import.meta.url)), 'receipts');
mkdirSync(receiptsDir, { recursive: true });

// ── JCS（RFC 8785）string escape ───────────────────────────────────
// 逐個 UTF-16 code unit escape：控制字符 + 所有非 ASCII → \uXXXX（小寫 hex）。
// 咁樣 emoji（surrogate pair）會出 \ud83d\ude80，同 Python ensure_ascii 一致。
function jcsEscape(s: string): string {
  let out = '';
  for (let i = 0; i < s.length; i++) {
    const code = s.charCodeAt(i);
    const ch = s[i];
    if (ch === '"') out += '\\"';
    else if (ch === '\\') out += '\\\\';
    else if (ch === '\n') out += '\\n';
    else if (ch === '\r') out += '\\r';
    else if (ch === '\t') out += '\\t';
    else if (ch === '\b') out += '\\b';
    else if (ch === '\f') out += '\\f';
    else if (code < 0x20) out += '\\u' + code.toString(16).padStart(4, '0');
    else if (code > 0x7e) out += '\\u' + code.toString(16).padStart(4, '0');
    else out += ch;
  }
  return out;
}

// ── canonicalJson：RFC 8785 JCS ────────────────────────────────────
export function canonicalJson(obj: unknown): string {
  if (obj === null) return 'null';
  if (typeof obj === 'boolean') return obj ? 'true' : 'false';
  if (typeof obj === 'number') return String(obj);
  if (typeof obj === 'bigint') return String(obj);
  if (typeof obj === 'string') return '"' + jcsEscape(obj) + '"';
  if (Array.isArray(obj)) return '[' + obj.map((x) => canonicalJson(x)).join(',') + ']';
  if (typeof obj === 'object') {
    // Object.keys().sort() 預設就係 UTF-16 code unit 排序
    const keys = Object.keys(obj as Record<string, unknown>).sort();
    return '{' + keys.map((k) =>
      '"' + jcsEscape(k) + '":' + canonicalJson((obj as Record<string, unknown>)[k])
    ).join(',') + '}';
  }
  throw new Error('canonicalJson: unsupported type ' + typeof obj);
}

// ── digest 工具 ─────────────────────────────────────────────────────
export function digestBytes(data: Buffer | Uint8Array): string {
  return 'sha256:' + createHash('sha256').update(data).digest('hex');
}

export function contentDigest(path: string): string {
  return digestBytes(readFileSync(path));
}

export function canonicalDigest(obj: unknown): string {
  return digestBytes(Buffer.from(canonicalJson(obj), 'utf8'));
}

export function canonicalDigestFile(path: string): string {
  const raw = readFileSync(path);
  try {
    const obj = JSON.parse(raw.toString('utf8'));
    if (obj === null || typeof obj !== 'object' || Array.isArray(obj)) {
      throw new Error('not a JSON object');
    }
    return canonicalDigest(obj);
  } catch {
    return digestBytes(raw); // 非 JSON object → fall back contentDigest
  }
}

// ── writeReceipt ──────────────────────────────────────────────────
export interface OutputMeta {
  content_digest: string;
  canonical_digest: string;
  status: string;
}

export interface Receipt {
  protocol: string;
  version: string;
  task_id: string;
  from: { agent_id: string; vendor: string };
  to: { agent_id: string; vendor: string };
  epoch: number;
  typed_reason: string;
  note: string;
  outputs: Record<string, OutputMeta>;
}

const SAFE_NAME_RE = /^[^/\\\x00]+$/;

function safeName(name: unknown): name is string {
  if (typeof name !== 'string' || name.length === 0) return false;
  if (name !== name.trim() || name.includes('\x00')) return false;
  if (name === '.' || name === '..' || name.includes('..')) return false;
  if (!SAFE_NAME_RE.test(name)) return false;
  return true;
}

const SAFE_TASK_ID_RE = /^[A-Za-z0-9_-]{1,128}$/;

export function writeReceipt(
  taskId: string,
  fromAgent: string,
  toAgent: string,
  prevOutputs: Array<string | { path: string; status?: string }>,
  newOutputs: Array<string | { path: string; status?: string }>,
  opts: { fromVendor?: string; toVendor?: string; typedReason?: string; note?: string } = {},
): Receipt {
  if (!SAFE_TASK_ID_RE.test(taskId)) {
    throw new Error('invalid task_id: ' + JSON.stringify(taskId));
  }
  const outputs: Record<string, OutputMeta> = {};
  let finalReason = opts.typedReason ?? PASS;

  for (const src of [prevOutputs, newOutputs]) {
    for (const item of src) {
      const pathStr = typeof item === 'string' ? item : (item.path || '');
      if (!pathStr) continue; // 空 path 略過
      if (!safeName(pathStr)) {
        finalReason = REJECT; // 唔安全 path fail-closed
        continue;
      }
      const p = join(receiptsDir, pathStr);
      if (!existsSync(p)) {
        finalReason = REJECT;
        outputs[pathStr] = { content_digest: digestBytes(Buffer.alloc(0)), canonical_digest: digestBytes(Buffer.alloc(0)), status: REJECT };
        continue;
      }
      const raw = readFileSync(p);
      if (raw.length === 0) {
        finalReason = REJECT; // 空輸出 = silent failure
        outputs[pathStr] = { content_digest: digestBytes(Buffer.alloc(0)), canonical_digest: digestBytes(Buffer.alloc(0)), status: REJECT };
        continue;
      }
      outputs[pathStr] = {
        content_digest: contentDigest(p),
        canonical_digest: canonicalDigestFile(p),
        status: PASS,
      };
    }
  }

  const receipt: Receipt = {
    protocol: 'rbp',
    version: '0.1',
    task_id: taskId,
    from: { agent_id: fromAgent, vendor: opts.fromVendor || fromAgent },
    to: { agent_id: toAgent, vendor: opts.toVendor || toAgent },
    epoch: Math.floor(Date.now() / 1000),
    typed_reason: finalReason,
    note: opts.note ?? '',
    outputs,
  };
  writeFileSync(join(receiptsDir, taskId + '.json'), canonicalJson(receipt), 'utf8');
  return receipt;
}

// ── densityCheck ──────────────────────────────────────────────────
export function densityCheck(receipt: Record<string, unknown>): [boolean, number] {
  let present = 0;
  for (const f of REQUIRED_FIELDS) {
    if (f in receipt && receipt[f] !== null && receipt[f] !== undefined) present++;
  }
  const metric = present / REQUIRED_FIELDS.length;
  if (metric < MIN_DENSITY) return [false, metric];
  const outputs = receipt.outputs;
  if (typeof outputs !== 'object' || outputs === null || Array.isArray(outputs) || Object.keys(outputs).length === 0) {
    return [false, 0];
  }
  return [true, metric];
}

// ── verifyReceipt ─────────────────────────────────────────────────
export function verifyReceipt(taskId: string, expectFrom: string, maxAge = 3600): [boolean, string] {
  const path = join(receiptsDir, taskId + '.json');
  if (!existsSync(path)) return [false, 'receipt file not found: ' + path];

  let receipt: unknown;
  try {
    receipt = JSON.parse(readFileSync(path, 'utf8'));
  } catch (e) {
    return [false, 'failed to read receipt: ' + String(e)];
  }
  if (receipt === null || typeof receipt !== 'object' || Array.isArray(receipt)) {
    return [false, 'receipt is not a JSON object'];
  }
  const r = receipt as Record<string, unknown>;

  // ① density
  const [dOk, density] = densityCheck(r);
  if (!dOk) return [false, 'INCOMPLETE: density ' + density.toFixed(2) + ' < ' + MIN_DENSITY];

  // ② scope
  const sender = r.from as Record<string, unknown> | undefined;
  if (typeof sender !== 'object' || sender === null || sender.agent_id !== expectFrom) {
    return [false, 'scope mismatch: receipt.from=' + JSON.stringify(sender) + ', expected=' + expectFrom];
  }

  // ③ digest
  const outputs = r.outputs as Record<string, OutputMeta> | undefined;
  if (typeof outputs !== 'object' || outputs === null) return [false, 'outputs is not a map'];
  for (const [name, meta] of Object.entries(outputs)) {
    if (!safeName(name)) return [false, 'unsafe output name: ' + JSON.stringify(name)];
    if (typeof meta !== 'object' || meta === null) return [false, 'output metadata not a map for ' + JSON.stringify(name)];
    const p = join(receiptsDir, name);
    if (!existsSync(p)) return [false, 'output file missing: ' + name];
    try {
      if (contentDigest(p) !== meta.content_digest) return [false, 'content_digest mismatch for ' + name];
      if (canonicalDigestFile(p) !== meta.canonical_digest) return [false, 'canonical_digest mismatch for ' + name];
    } catch (e) {
      return [false, 'digest failed for ' + name + ': ' + String(e)];
    }
  }

  // ④ epoch
  const epoch = r.epoch as number;
  if (typeof epoch !== 'number' || !Number.isInteger(epoch)) {
    return [false, 'invalid epoch type: ' + typeof epoch];
  }
  const age = Math.floor(Date.now() / 1000) - epoch;
  if (age < 0 || age > maxAge) return [false, 'epoch invalid/stale: age=' + age + 's (max_age=' + maxAge + 's)'];

  // ⑤ verdict
  if (r.typed_reason !== PASS) return [false, 'verdict not PASS: ' + String(r.typed_reason)];

  return [true, 'OK'];
}
